Assets/MixedRealityToolkit-SDK/Features/UX/Interactable/Scripts/Events/OnClickReceiver.cs

```diff
@@ -0,0 +1,51 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Events;
+
+namespace Microsoft.MixedReality.Toolkit.SDK.UX
+{
+    public class OnClickReceiver : ReceiverBase
+    {
+        [InspectorField(Type = InspectorField.FieldTypes.Float, Label = "Click Time", Tooltip = "The press and release should happen within this time")]
+        public float ClickTime = 0.5f;
+
+        private float clickTimer = 0;
+
+        private bool hasDown;
+        private State lastState;
+
+        public OnClickReceiver(UnityEvent ev): base(ev)
+        {
+            Name = "OnClick";
+        }
+
+        public override void OnUpdate(InteractableStates state, Interactable source)
+        {
+            bool hadDown = hasDown;
+            hasDown = state.GetState(InteractableStates.InteractableStateEnum.Pressed).Value > 0;
+
+            bool focused = state.GetState(InteractableStates.InteractableStateEnum.Focus).Value > 0;
+
+            if (hadDown && !hasDown && focused && clickTimer < ClickTime)
+            {
+                Debug.Log("Click!");
+                uEvent.Invoke();
```

Comment:
Not sure we should be using Unity Events. We've got a pretty solid event system we're already using that doesn't require uGUI.

Reply:
The UnityEvents are for designers to use in editor, unless we have an event system with a property drawer. I'm happy to change how the events work, but the main point is that designers can hook them up without code.

Reply:
>the main point is that designers can hook them up without code.

Def agree! Let's sync and discuss how we can accommodate this.

My other goal as well is to try and not be dependent on unity gui as well. That's mainly for 2d screens and wasn't really designed with 3d in mind.

Reply:
Oh, I thought that because we were using unity events that they it was strictly uGUI based. In that case, we can def try to stick to the handlers we've setup (very similar to old HTK input). We should still be able to setup events in the inspector without any code. (take a look at PointerClickHandler and SpeechInputHandler classes as examples). We can still use unity events in a small capacity where needed, but I'd like to make sure we're just exposing the fields we need to reference when handling events.

Reply:
Take a look at [my comments from this issue related to Unity Events](https://github.com/Microsoft/MixedRealityToolkit-Unity/issues/876#issuecomment-323417651)

Reply:
For sure, I expose the basic click event because if follows the patterns on button, but there are a ton of other events that the designer can add as needed and they are very specific. In the last version of InteractiveElements I thought this button could be used in 5 ways, so I exposed 5 UnityEvents. That just makes things complicated and quite frankly it was easier to building custom inspectors. Now only one event is shown at default and if the design wants another event, they can add that specific event. The same goes for an engineer that would not like to use the inspector, but stick to code, they can target a pattern that defines an event. This part still needs some work though. I know there's a ton here. Thanks for taking the time.

Reply:
I def love where this is going, and there's a ton of awesome work you (and I'm sure others) have put into this. I'd like to dig in a bit more before I add any additional feedback.